### PR TITLE
azurerm_frontdoor_custom_https_configuration key vault certificate source: do not require version

### PR DESCRIPTION
**Pull request by rubbish**
if using azure key vault as source of certificate, then the secret version is not required. this removes that validation.

I still need to update documentation and perhaps tests. But reading over https://github.com/terraform-providers/terraform-provider-azurerm/issues/10935 I'm not sure how this validation got added.

Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/10935

### CODE REVIEW

**Review by WodansSon — CHANGES_REQUESTED** [review]
@rubbish thanks for the PR and this is looking really good so far. There are few thinks I would like to see before we can merge this if you don't mind. 

1) Could we add a test case for this condition?
2) If you can mention this in the documentation, that would be great!

**Comment by rubbish** [comment]
I've updated the docs and added a test, but I cannot get the test to run. I get currently:

```
+ make acctests SERVICE=frontdoor TESTARGS=-run=TestAccFrontDoorCustomHttpsConfiguration_CustomHttps TESTTIMEOUT=60m
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/frontdoor -run=TestAccFrontDoorCustomHttpsConfiguration_CustomHttps -timeout 60m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/04/14 17:04:20 [DEBUG] not using binary driver name, it's no longer needed
2021/04/14 17:04:20 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== PAUSE TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
    testing.go:620: Step 1/2 error: Error running pre-apply refresh: 2021/04/14 17:04:27 [DEBUG] Using modified User-Agent: Terraform/0.12.21 HashiCorp-terraform-exec/0.10.0

        Error: Could not satisfy plugin requirements


        Plugin reinitialization required. Please run "terraform init".

        Plugins are external binaries that Terraform uses to access and manipulate
        resources. The configuration provided requires plugins which can't be located,
        don't satisfy the version constraints, or are otherwise incompatible.

        Terraform automatically discovers provider requirements from your
        configuration, including providers used in child modules. To see the
        requirements and constraints from each module, run "terraform providers".



        Error: provider.azurerm: no suitable version installed
          version requirements: "(any version)"
          versions installed: none


--- FAIL: TestAccFrontDoorCustomHttpsConfiguration_CustomHttps (7.06s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor   8.919s
FAIL
make: *** [acctests] Error 1

```

Any idea on what to try @WodansSon ?

**Comment by WodansSon** [comment]
@rubbish, what directory are you running the tests from? You should be in the root folder of your repo.

**Review by katbyte — CHANGES_REQUESTED** [review]
When i try to run the tests i get:
```
------- Stdout: -------
=== RUN   TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== PAUSE TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
    testing.go:620: Step 3/3 error: Error running pre-apply refresh: 
        Error: Missing required argument
        
          on config535635520/terraform_plugin_test.tf line 78, in resource "azurerm_key_vault" "test":
          78: resource "azurerm_key_vault" "test" {
        
        The argument "sku_name" is required, but no definition was found.
        
        
        Error: Unsupported argument
        
          on config535635520/terraform_plugin_test.tf line 82, in resource "azurerm_key_vault" "test":
          82:   sku                 = "Standard"
        
        An argument named "sku" is not expected here.
        
    testing_new.go:21: WARNING: destroy failed, so remote objects may still exist and be subject to billing
    testing_new.go:21: failed to destroy: 
        Error: Missing required argument
        
          on config535635520/terraform_plugin_test.tf line 78, in resource "azurerm_key_vault" "test":
          78: resource "azurerm_key_vault" "test" {
        
        The argument "sku_name" is required, but no definition was found.
        
        
        Error: Unsupported argument
        
          on config535635520/terraform_plugin_test.tf line 82, in resource "azurerm_key_vault" "test":
          82:   sku                 = "Standard"
        
        An argument named "sku" is not expected here.
        
        
--- FAIL: TestAccFrontDoorCustomHttpsConfiguration_CustomHttps (1870.34s)
FAIL

------- Stderr: -------
2021/04/15 03:11:28 [DEBUG] not using binary driver name, it's no longer needed
2021/04/15 03:11:28 [DEBUG] not using binary driver name, it's no longer needed
```

**Comment by rubbish** [comment]
@WodansSon I was in the root directory of the project. I did not have it cloned under my gopath however. But that didn't make any difference.

@katbyte My bad, I've fixed that now and pushed it up.

File: website/docs/r/frontdoor_custom_https_configuration.html.markdown
Comment (WodansSon) [inline website/docs/r/frontdoor_custom_https_configuration.html.markdown:119]:
```suggestion
* `azure_key_vault_certificate_secret_version` - (Optional) The version of the Key Vault secret representing the full certificate PFX. Defaults to `Latest`.
```

**Review by WodansSon — CHANGES_REQUESTED** [review]
@rubbish, thanks for pushing those changes. I made a very minor comment on the documentation, once you get that updated and if the test pass we can get this merged. 🚀

**Comment by WodansSon** [comment]
@rubbish, the test is failing with:
```
=== RUN   TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== PAUSE TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
testing.go:620: Step 3/3 error: Error running pre-apply refresh:
Error: "name" may only contain alphanumeric characters and dashes and must be between 3-24 chars
on config982909111/terraform_plugin_test.tf line 77, in resource "azurerm_key_vault" "test":
77: resource "azurerm_key_vault" "test" {
testing_new.go:21: WARNING: destroy failed, so remote objects may still exist and be subject to billing
testing_new.go:21: failed to destroy:
Error: "name" may only contain alphanumeric characters and dashes and must be between 3-24 chars
on config982909111/terraform_plugin_test.tf line 77, in resource "azurerm_key_vault" "test":
77: resource "azurerm_key_vault" "test" {
--- FAIL: TestAccFrontDoorCustomHttpsConfiguration_CustomHttps (1095.10s)
FAIL
``` 

Looks like you are going to have to shorten the name of the key vault in the test, maybe something like this?
```
name = "vault%d"
```

File: website/docs/r/frontdoor_custom_https_configuration.html.markdown
Comment (timja) [reply website/docs/r/frontdoor_custom_https_configuration.html.markdown:119]: may be too long, but what about:
```suggestion
* `azure_key_vault_certificate_secret_version` - (Optional) The version of the Key Vault secret. Defaults to 'Latest'. This will automatically be updated by Azure as new versions become available in the Key Vault, (it may take up to 24 hours to update).
```

Comment (rubbish) [reply website/docs/r/frontdoor_custom_https_configuration.html.markdown]: I went with @WodansSon's suggestion. IMO the Azure documentation should specify the Latest behavior, not necessarily the terraform documentation.

**Review by WodansSon — APPROVED** [review]
@rubbish, thanks for pushing those changes. This LGTM now. 🚀

**Comment by WodansSon** [comment]
@rubbish, the test is still failing with:
```
=== RUN   TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== PAUSE TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
testing.go:620: Step 3/3 error: Error running pre-apply refresh:
Error: validating Front Door Custom Https Configuration for Endpoint "frontend-endpoint-custom" (Front Door "acctest-FD-210420230753384872" / Resource Group "acctestRG-frontdoor-210420230753384872"): "frontend_endpoint":"/subscriptions/*******/resourceGroups/acctestRG-frontdoor-210420230753384872/providers/Microsoft.Network/frontDoors/acctest-FD-210420230753384872/frontendEndpoints/frontend-endpoint-custom" "custom_https_configuration" is invalid, all of the following keys must have values in the "custom_https_configuration" block: "azure_key_vault_certificate_secret_name" and "azure_key_vault_certificate_vault_id"
on config431380062/terraform_plugin_test.tf line 66, in resource "azurerm_frontdoor_custom_https_configuration" "test":
```

**Comment by rubbish** [comment]
@WodansSon  So I've sort of hit a road block with the test. Frontdoor requires that certificate not to be self-signed:

```
Message="The certificate imported from the Key Vault is a Self Signed certificate and is not permitted for Bring Your Own Certificate"
```

I'm not sure what to do. I think we might have to do this change without a test. Thoughts?

**Comment by timja** [comment]
> @WodansSon So I've sort of hit a road block with the test. Frontdoor requires that certificate not to be self-signed:
> 
> ```
> Message="The certificate imported from the Key Vault is a Self Signed certificate and is not permitted for Bring Your Own Certificate"
> ```
> 
> I'm not sure what to do. I think we might have to do this change without a test. Thoughts?

I thought this functionality was intentionally not tested in the provider because of this requirement?
There's been previous discussion about it but I can't find it right now.

I would back out the test addition @rubbish

**Comment by WodansSon** [comment]
@rubbish, I would tend to agree with @timja go ahead and back the test out, but if you can do a manual test pass on it with a versionless key and a versioned key I think that would be sufficient. Thanks again! 🚀

**Comment by rubbish** [comment]
@WodansSon @timja That test is now removed. I did test it out manually locally using an issued certificate in key vault with and without a version. It works as intended.

**Comment by joaomoreno** [comment]
This bot is so kind, thank you!

**Pull request merged**
